examples/fabric/build_your_own_trainer/trainer.py

```diff
@@ -0,0 +1,522 @@
+            grad_accum_steps: How many batches to process before each optimizer step
+            limit_train_batches: Limits the number of train batches per epoch
+                If greater than number of batches in the dataloader, this has no effect.
+            limit_val_batches: Limits the number of validation batches per epoch.
+                If greater than number of batches in the dataloader, this has no effect.
+            validation_frequency: How many epochs to run before each validation epoch.
+            use_distributed_sampler: Wraps the sampler of each dataloader with a respective distributed-aware sampler
+                in case of distributed training.
+            checkpoint_dir: Directory to store checkpoints to.
+            checkpoint_frequency: How many epochs to run before each checkpoint is written.
+
+        Warning:
+            callbacks written for the lightning trainer (especially making assumptions on the trainer), won't work!
+        """
+
+        self.fabric = Fabric(
+            accelerator=accelerator,
+            strategy=strategy,
+            devices=devices,
+            precision=precision,
+            plugins=plugins,
+            callbacks=callbacks,
+            loggers=loggers,
+        )
+        self.global_step = 0
+        self.grad_accum_steps: int = grad_accum_steps
+        self.current_epoch = 0
+
+        self.max_epochs = max_epochs
+        self.max_steps = max_steps
+        self.should_stop = False
+
+        # ensures limit_X_batches is either int or inf
+        if not isinstance(limit_train_batches, int):
+            assert limit_train_batches == float("inf")
+
+        if not isinstance(limit_val_batches, int):
+            assert limit_val_batches == float("inf")
+
+        self.limit_train_batches = limit_train_batches
+        self.limit_val_batches = limit_val_batches
+        self.validation_frequency = validation_frequency
+        self.use_distributed_sampler = use_distributed_sampler
+        self._current_train_return: Union[torch.Tensor, Mapping[str, Any]] = {}
+        self._current_val_return: Optional[Union[torch.Tensor, Mapping[str, Any]]] = {}
+
+        self.checkpoint_dir = checkpoint_dir
+        self.checkpoint_frequency = checkpoint_frequency
+
+    def fit(
+        self,
+        model: L.LightningModule,
+        train_loader: torch.utils.data.DataLoader,
+        val_loader: torch.utils.data.DataLoader,
+    ):
+        """The main entrypoint of the trainer, triggering the actual training.
+
+        Args:
+            model: the LightningModule to train.
+                Can have the same hooks as :attr:`callbacks` (see :meth:`MyCustomTrainer.__init__`).
+            train_loader: the training dataloader. Has to be an iterable returning batches.
+            val_loader: the validation dataloader. Has to be an iterable returning batches.
+                If not specified, no validation will run.
+        """
+        self.fabric.launch()
```

Comment:
What happens if you call `fit` twice L.Trainer style? Does the second `.launch()` become a no-op? cc @awaelchli

Reply:
Yes, it does nothing

Reply:
I'd prefer to only launch when actually fitting